apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts

```diff
@@ -1,30 +1,33 @@
 import { Injectable } from '@nestjs/common';
-import { JobStatusEnum, JobRepository } from '@novu/dal';
+import { JobStatusEnum, JobRepository, JobEntity } from '@novu/dal';
+import { StepTypeEnum } from '@novu/shared';
 import { CancelDelayedCommand } from './cancel-delayed.command';
+import { isMainDigest } from '@novu/application-generic/build/main/utils/digest';
 
+type PartialJob = Pick<JobEntity, '_id' | 'type' | 'status' | '_environmentId' | '_subscriberId'>;
 @Injectable()
 export class CancelDelayed {
   constructor(private jobRepository: JobRepository) {}
 
   public async execute(command: CancelDelayedCommand): Promise<boolean> {
-    const jobs = await this.jobRepository.find(
+    const transactionJobs: PartialJob[] = await this.jobRepository.find(
       {
         _environmentId: command.environmentId,
         transactionId: command.transactionId,
-        status: JobStatusEnum.DELAYED,
+        status: [JobStatusEnum.DELAYED, JobStatusEnum.MERGED, JobStatusEnum.PENDING],
```

Comment:
JobStatusEnum.MERGED was added by David in order to solve the 2nd, 3rd, etc.. cancellations.

In addition, I added JobStatusEnum.PENDING as it looks like we were keeping transaction child pending jobs in the database that would never resolve. IMO they should be canceled as well, let me know if someone knows if it was intentional.

Reply:
If I remember correctly we skipped pending because there could be a matter of who getting first in workflows without digest. Like what happens if they cancel it and we find it here at the same time it is picked up and put on the queue. How will it be handled?

Reply:
I see, so i think about two different points we need to think about here:
1. this should be okay because, at the current project state, we do not support cancelation of a job of CHANNEL type, meaning only ACTION which is delay and digest supports cancelation.
2. we do not support Channel cancelation, meaning if we cancel here workflow without digest we will get the following flow with corrupted status:
    1. run a job (in-app for example), 
    2. which will be executed because we won't check its status (cancel)
    3. once we finish the execution we mark it as Completed.
    

Let me know if you think I missed something in the flow.

One solution would be to add another query after
```
.find({
      {
        _environmentId: command.environmentId,
        transactionId: command.transactionId,
        status: [JobStatusEnum.DELAYED, JobStatusEnum.MERGED,
        })
```
if transactionJobs is truthy it means that this is a digest workflow transaction fetch for its PENDING jobs and then cancel them all.

Reply:
I think we can go with your solution because cancel and then complete it could create confusion for the user.

Reply:
Awesome thanks for the feedback we could have missed it otherwise!! i will create an update :)